vRO Knative VEBA Function

dsbibby opened the pull request:
Signed-off-by: David Bibby <davidsbibby@gmail.com>

## Summary

This is a Knative function for VEBA that calls a vRO workflow, passing all Cloud Event parameters to the workflow as input variables. It also converts vCenter SDK objects from the Cloud Event in to native vRO objects.

Marked as WIP as documentation still needs creating but this was asked about on the Slack channel.

## Pull Request Checklist
🚨 Please review the [guidelines for contributing](https://vmweventbroker.io/community) to this repository.

- [x] Please ensure that you are making a pull request against the **Development** branch
- [x] Please use the `WIP` keyword in the title of your PR if you are not ready for review
- [ ] Please ensure that you have opened a Github Issue if you are resolving/fixing a problem
- [x] Please ensure that you have [signed](https://help.github.com/en/github/authenticating-to-github/signing-commits) all commits and that you have [squashed](https://medium.com/@slamflipstrom/a-beginners-guide-to-squashing-commits-with-git-rebase-8185cf6e62ec) all relevant commits related to your change
- [x] Please make sure that you have tested your change locally by successfully [building and deploying the VMware Event Broker Appliance](https://vmweventbroker.io/kb/contribute-appliance) and/or [building and deploying VMware Event Router](https://vmweventbroker.io/kb/contribute-eventrouter)
- [ ] Please include any relevant screenshots and/or output as part of your testing
- [ ] Please include any documentation updates that is applicable for your changes

## Change Type

What types of changes does your code introduce to the VMware Event Broker Appliance?

_Put an `x` in all boxes that apply_

Please check the type of change your PR introduces:
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation changes
- [x] Other (please describe): New Example Function

## Resolved Issues

List of Issues closed or resolved by this PR. Add multiple `Closes` keyword followed by the issue number (e.g. Closes #ISSUE-NUMBER)

Closes #ISSUE-NUMBER

## Testing Verification

* Short summary of testing (e.g. successfully built and deployed VMware Event Broker Appliance)
* Please include any relevant screenshots and/or output as part of the testing

## Additional Information

* Any other details you wish to include or mention

If you have any questions/comments, feel free to reach out to team on Slack [#vcenter-event-broker-appliance](https://vmwarecode.slack.com/archives/CQLT9B5AA)

Thank you from the VEBA Team! 🥳

embano1 commented on examples/knative/python/kn-py-vro/kn-py-vro_secret.json:
```diff
@@ -0,0 +1,10 @@
+{
+    "SERVER": "VRO_FQDN_OR_IP",
+    "PORT": 443,
+    "AUTH_TYPE": "BASIC/OAUTH"
```
nit: trailing `,` missing.

dsbibby replied:
I spotted that in my local version with real creds in but didn't fix the commited version. Will sort.

embano1 commented on examples/knative/python/kn-py-vro/handler.py:
```diff
@@ -0,0 +1,270 @@
+    body = {"parameters": []}
+    for item in event.items():
+        # Get the vRO input parameter correctly formatted for the received event data
+        res = getVroInputParam(item)
```
question: IMHO `getVroInputParam` can throw an exception in certain cases (e.g. dict lookup). Should we add try/catch around these calls?

dsbibby replied:
Agree that there's at least a couple of places in `getVroInputParam` where I've assumed the format of the passed data. If you agree I feel it would be better to add the exception handling in `getVroInputParam` itself.

dsbibby replied:
After reviewing the code in the `getVroInputParam` function I decided that I don't think exception handling is required as the couple of places I could see that could cause an exception could be coded a little more defensively.

Let me know what you think of it as it stands to ensure I've not missed anything. My aim here is that any cloud event parameter makes it in to the vRO request somehow, even if it just ends up as a string of JSON as at least then the vRO workflow could handle this.

embano1 replied:
I'm not quite sure I follow why the conversion from CE.date to event and corresponding CE attributes is needed. Same for `event.items()`. I'll wait until you get some unit tests in to see how the code behaves under various conditions.

dsbibby replied:
My aim here is to parse each CE.data item and each CE attribute in to vRO input parameters - one for each so they can be picked up in vRO. At the very least I'll change the naming of the `event` variable here as that's not descriptive.

embano1 commented on examples/knative/python/kn-py-vro/handler.py:
```diff
@@ -0,0 +1,270 @@
+    cevent = from_http(request.headers, request.get_data(),None)
+
+    if cevent._attributes.get("datacontenttype").lower() != "application/json":
+            return get_response(f'invalid datacontenttype for cloud event: {cevent._attributes.get("datacontenttype")}', 500)
```
nit: invalid data should always return `400`

embano1 commented on examples/knative/python/kn-py-vro/handler.py:
```diff
@@ -0,0 +1,270 @@
+    app.logger.debug(f'{bgc.HEADER}Reading Cloud Event: {bgc.ENDC}')
+    cevent = from_http(request.headers, request.get_data(),None)
+
+    if cevent._attributes.get("datacontenttype").lower() != "application/json":
```
nit: I think for CE attributes you can directly use dict syntax and not internal methods

embano1 commented on examples/knative/python/kn-py-vro/handler.py:
```diff
@@ -0,0 +1,270 @@
+        # CloudEvent - simple validation
+        source = cevent._attributes['source']
+        event = cevent.data
+        _ = event['Key']
```
question: is this needed?

dsbibby replied:
I did this as a crude way to confirm the presence of the 'Key' parameter in the event data (i.e. is this valid CloudEvent data and it's all parsed properly, and is a hang up from the OpenFaaS version of the function. I think I'm right in thinking that with Knative and  the CloudEvent datatype, *and* the tests done previous to this that this is no longer serving any useful purpose.

embano1 replied:
Yup, shouldn't be needed.

embano1 commented on examples/knative/python/kn-py-vro/handler.py:
```diff
@@ -0,0 +1,270 @@
+        return get_response(f'Invalid JSON, data not iterable > TypeError: {err}', 500)
+
+    app.logger.debug(f'{bgc.HEADER}Validating Input data: {bgc.ENDC}')
+    app.logger.debug(f'{bgc.OKBLUE}Event > {bgc.ENDC}{json.dumps(event, indent=4, sort_keys=True)}')
```
question: do we want to dump via json or just use the object's string representation? (I think the class implements str IIRC)

dsbibby replied:
That is correct, although in terms of readability in the log output, the string representation of CloudEvent is an unformatted single line, whereas this line makes it easier to read. TBH this is probably only useful during development of the function, so I'm happy either way.

embano1 replied:
Ok, if the `json.dumps` output does not mess up the object/content and makes it easier to read, go for it!

embano1 commented on examples/knative/python/kn-py-vro/handler.py:
```diff
@@ -0,0 +1,270 @@
+    app.logger.debug(f'{bgc.OKBLUE}Event > {bgc.ENDC}{json.dumps(event, indent=4, sort_keys=True)}')
+
+    vchost = urllib3.util.url.parse_url(source).host
+    app.logger.debug(f'{bgc.HEADER}All tests passed! Build vRO request:{bgc.ENDC}')
```
nit: tests -> validation

embano1 commented on examples/knative/python/kn-py-vro/handler.py:
```diff
@@ -0,0 +1,270 @@
+        # Surpress SSL warnings
+        urllib3.disable_warnings(urllib3.exceptions.InsecureRequestWarning)
+    
+    get_token()
```
question: what if this throws?

dsbibby replied:
Yep, exception handling incoming...!

dsbibby replied:
This took some work to get it to work properly within the Flask framework, but hopefully is all OK now. Better handling added around the reading of the config from the secret too.

embano1 commented:
Thx! Are you also thinking of adding some unit tests around the input event handling and outgoing request logic?

embano1 commented on examples/knative/python/kn-py-vro/kn-py-vro_secret.json:
```diff
@@ -0,0 +1,10 @@
+    "USERNAME": "VRO_USERNAME",
+    "PASSWORD": "VRO_PASSWORD",
+    "REFRESH_TOKEN": "VRAC_ONLY",
+    "INSECURE_SSL": true,
```
nit: default should always be safe, i.e. `false`

embano1 commented:
Also, there seem to be some openfaas related files in some commits.

dsbibby commented:
> Also, there seem to be some openfaas related files in some commits.

Oops! I shall tidy that up 👍 


> Thx! Are you also thinking of adding some unit tests around the input event handling and outgoing request logic?

Yes, that would be a good thing to add... I'll have a think about the best approach to this. It would be simple enough to test that the outgoing request to vRO is valid for certain passed input events, but it would be nice to also confirm that a vRO workflow correctly receives the data etc. I have some ideas I'll bash about a bit.

dsbibby commented:
I think I've address most of your previous comments - let me know your thoughts.

I've not yet done any work on unit testing but I have some ideas around using a vRO workflow to help. I'll work on this.

embano1 commented on examples/knative/python/kn-py-vro/README.md:
```diff
@@ -0,0 +1,272 @@
+
+The function code not only demonstrates a number of techniques (such as creating persistent OAUTH sessions, and the vRO workflow API), but is generic enough to be usable in a multitude of scenario's out-of-the-box.
+
+> **Note:** Only Cloud Event payload ("*data*") in JSON format is supported.
```
```suggestion
> **Note:** Only CloudEvent payload (`data`) in JSON format is supported.
```

embano1 commented on examples/knative/python/kn-py-vro/README.md:
````diff
@@ -0,0 +1,272 @@
+- Docker
+
+```bash
+export IMAGE=<docker-username>/kn-py-vro
+pack build --builder gcr.io/buildpacks/builder:v1 ${IMAGE}
````
```suggestion
export IMAGE=<docker-username>/kn-py-vro:1.0
pack build --builder gcr.io/buildpacks/builder:v1 ${IMAGE}
```

embano1 commented on examples/knative/python/kn-py-vro/README.md:
````diff
@@ -0,0 +1,272 @@
+2. Run a local container based on the `${IMAGE}` image or the one you just built.
+
+    ```bash
+    docker run -e PORT=8080 -it --rm -p 8080:8080 --env VCCONFIG_SECRET="$(cat kn-py-vm-attr_secret.json)" ${IMAGE}
````
nit: would that work on windows (`cat`)?

dsbibby replied:
"Who uses windows!?" - oh wait, me right now! I'll fix.

I based this on the kn-py-vm-attr function - we should fix a few bits in there too. Should I raise an issue?

embano1 commented on examples/knative/python/kn-py-vro/README.md:
```diff
@@ -0,0 +1,272 @@
+You need:
+
+* A deployed VEBA instance (K8s or appliance based)
+* A deployed Knative setup
```
nit: redundant since VEBA has Knative

embano1 commented on examples/knative/python/kn-py-vro/function.yaml:
```diff
@@ -0,0 +1,36 @@
+        autoscaling.knative.dev/minScale: "1"
+    spec:
+      containers:
+        - image: pilue/kn-py-vro:1.0
```
```suggestion
        - image: us.gcr.io/daisy-284300/veba/kn-py-vro:1.0
```

embano1 commented on examples/openfaas/python/echo/stack.yml:
```diff
@@ -1,7 +1,7 @@
 version: 1.0
```
nit: obsolet

dsbibby replied:
Gosh - had no idea I'd committed that! Will remove.

dsbibby replied:
Note that this is actually me accidentally push my local changes to the OpenFaaS echo function. I've reverted these changes to the upstream/development version, so this should be ignored when I squash I think.

embano1 commented on examples/knative/python/kn-py-vro/handler.py:
```diff
@@ -0,0 +1,325 @@
+
+    # Load the Events that function gets from vCenter through the Event Router
+    app.logger.debug(f'{bgc.HEADER}Reading Cloud Event{bgc.ENDC}')
+    cevent = from_http(request.headers, request.get_data(),None)
```
question: could this throw?

dsbibby replied:
Yes it could - and it has in my testing when I've done something stupid! I'll catch this.

dsbibby commented:
I think I've resolved all the review comments so far with the exception of unit testing - I'm working on this still.

Many thanks!

embano1 commented:
@dsbibby do you think you can finish this PR until end of this week? Would be good to get this in for `v0.7.1` which we want to cut a pre-release early next week 🙏

dsbibby commented:
@embano1 - I'd like to hope so... although I've used this as an excuse to learn `pytest` so have been dealing with that learning curve at the same time!

I'll commit wherever I've got to later on this week/weekend and we can see where we're at. Many thanks 😃

dsbibby commented:
@embano1 - I've added a stack of unit tests now. Take a look and see what you think. I'd still like to get a more structured functional test in, but that could always come later if I don't get time this week. Many thanks!

embano1 reviewed (commented):
Fantastic work! Can you please squash and merge into one commit following our title/body structure (example below)?

```console
git commit -s -m "feat: Add Knative vRO function" -m "Closes: #728"
```

dsbibby commented:
@embano1 - all done I think. Added an extra unit test around the camel case function as I spotted a bug with it. Fixed the bug too of course!

Many thanks for the review 😄

embano1 reviewed (approved):
Massive work, thx a ton!

dsbibby commented:
Just though @embano1, I didn't include the relevant web site updates to link to this function. Shall I submit a separate PR for this?

embano1 commented:
Yes, there's a tracking issue for this: https://github.com/vmware-samples/vcenter-event-broker-appliance/issues/730

You'd do us a HUGE favor if you could add website entries for all the new Knative functions.
